File: web/src/api/workers.ts
import { apiUrl, httpClient } from "./client.ts";

export interface Worker {
  id: string;
  name: string;
  description: string;
  created_at: string;
  updated_at: string;
  last_seen: string | null;
  config: Record<string, unknown>;
  status: string;
}

export interface WorkerCreateResponse {
  worker: Worker;
  token: string;
}

export interface WorkerListResponse {
  workers: Worker[];
  total: number;
}

export interface WorkerTaskQueueItem {
  id: string;
  task_type: string;
  content_kind?: string | null;
  content_item_id?: string | null;
  folder_uuid: string;
  relative_path: string;
  status: string;
  stage?: string | null;
  requested_by_sub?: string | null;
  claimed_by?: string | null;
  claimed_at?: string | null;
  claim_age_seconds?: number | null;
  stale_after_seconds: number;
  is_stale: boolean;
  retry_count: number;
  max_retries: number;
  error_message?: string | null;
  created_at: string;
  updated_at: string;
}

export interface WorkerTaskQueueListResponse {
  tasks: WorkerTaskQueueItem[];
  total: number;
}

export interface WorkerTaskQueueCleanupResponse {
  total: number;
  requeued: number;
  failed: number;
}

export interface WorkerInstallPlatform {
  id: string;
  label: string;
  kind: "pip" | "docker";
  extra: string | null;
  extra_index_url: string | null;
  image: string | null;
  gpu: boolean;
  notes: string | null;
}

export interface WorkerInstallInfo {
  package: string;
  version: string;
  default_capabilities: string;
  public_url: string | null;
  platforms: WorkerInstallPlatform[];
}

export const workersApi = {
  list: async (): Promise<WorkerListResponse> => {
    const { json } = await httpClient(`${apiUrl}/workers/`);
    return json;
  },

  listTasks: async (
    options: { activeOnly?: boolean; limit?: number } = {},
  ): Promise<WorkerTaskQueueListResponse> => {
    const params = new URLSearchParams();
    params.set("active_only", String(options.activeOnly ?? true));
    if (options.limit) params.set("limit", String(options.limit));
    const { json } = await httpClient(`${apiUrl}/workers/tasks?${params}`);
    return json;
  },

  cleanupStaleTasks: async (): Promise<WorkerTaskQueueCleanupResponse> => {
    const { json } = await httpClient(`${apiUrl}/workers/tasks/cleanup-stale`, {
      method: "POST",
    });
    return json;
  },

  get: async (id: string): Promise<Worker> => {
    const { json } = await httpClient(`${apiUrl}/workers/${id}`);
    return json;
  },

  create: async (data: { name: string; description?: string }): Promise<WorkerCreateResponse> => {
    const { json } = await httpClient(`${apiUrl}/workers/`, {
      method: "POST",
      body: JSON.stringify(data),
    });
    return json;
  },

  update: async (id: string, data: { name?: string; description?: string }): Promise<Worker> => {
    const { json } = await httpClient(`${apiUrl}/workers/${id}`, {
      method: "PUT",
      body: JSON.stringify(data),
    });
    return json;
  },

  delete: async (id: string): Promise<void> => {
    await httpClient(`${apiUrl}/workers/${id}`, {
      method: "DELETE",
    });
  },

  rotateToken: async (id: string): Promise<WorkerCreateResponse> => {
    const { json } = await httpClient(`${apiUrl}/workers/${id}/rotate-token`, {
      method: "POST",
    });
    return json;
  },

  installInfo: async (): Promise<WorkerInstallInfo> => {
    const { json } = await httpClient(`${apiUrl}/workers/install-info`);
    return json;
  },
};

File: api/routers/workers.py
"""Worker management API router."""

import logging
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.exc import IntegrityError

from auth.dependencies import require_admin
from models.user import User
from models.worker import (
    WorkerCreate,
    WorkerUpdate,
    WorkerResponse,
    WorkerCreateResponse,
    WorkerInstallInfo,
    WorkerInstallPlatform,
    WorkerListResponse,
    WorkerTaskQueueCleanupResponse,
    WorkerTaskQueueListResponse,
)
from services.worker import WorkerService
from services.general_settings import GeneralSettingsService
from database import get_db
from rls import set_rls_context, user_context
from version import get_app_version

router = APIRouter()
logger = logging.getLogger(__name__)

# Default capabilities a worker is started with (mirrors the compose default).
_DEFAULT_WORKER_CAPABILITIES = "document,video,image"

# GHCR namespace for the prebuilt worker images (REGISTRY/github.repository).
_WORKER_IMAGE_BASE = "ghcr.io/intextum/intextum"

_PYTORCH_CPU_INDEX = "https://download.pytorch.org/whl/cpu"
_PYTORCH_CUDA_INDEX = "https://download.pytorch.org/whl/cu126"

# Install targets for the Add-Worker onboarding flow. macOS Torch wheels are on
# PyPI (no extra index); Linux/Windows CPU/CUDA builds come from the PyTorch
# index; Docker targets use the prebuilt GHCR images.
_WORKER_INSTALL_PLATFORMS: list[WorkerInstallPlatform] = [
    WorkerInstallPlatform(
        id="macos-mps",
        label="macOS (Apple Silicon · MPS)",
        kind="pip",
        extra="mps",
        notes="Uses Apple's native Vision OCR (ocrmac) and MPS acceleration.",
    ),
    WorkerInstallPlatform(
        id="linux-cpu",
        label="Linux (CPU)",
        kind="pip",
        extra="cpu",
        extra_index_url=_PYTORCH_CPU_INDEX,
    ),
    WorkerInstallPlatform(
        id="linux-cuda",
        label="Linux (NVIDIA CUDA 12.6)",
        kind="pip",
        extra="cuda",
        extra_index_url=_PYTORCH_CUDA_INDEX,
    ),
    WorkerInstallPlatform(
        id="windows-cpu",
        label="Windows (CPU)",
        kind="pip",
        extra="cpu",
        extra_index_url=_PYTORCH_CPU_INDEX,
    ),
    WorkerInstallPlatform(
        id="windows-cuda",
        label="Windows (NVIDIA CUDA 12.6)",
        kind="pip",
        extra="cuda",
        extra_index_url=_PYTORCH_CUDA_INDEX,
    ),
    WorkerInstallPlatform(
        id="docker-cpu",
        label="Docker (CPU)",
        kind="docker",
        image=f"{_WORKER_IMAGE_BASE}/worker-cpu",
    ),
    WorkerInstallPlatform(
        id="docker-cuda",
        label="Docker (NVIDIA CUDA)",
        kind="docker",
        image=f"{_WORKER_IMAGE_BASE}/worker-cuda",
        gpu=True,
    ),
]


async def get_worker_service(
    user: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> WorkerService:
    await set_rls_context(db, user_context(user))
    return WorkerService(db=db)


def _is_unique_constraint_error(exc: IntegrityError) -> bool:
    """Best-effort detection for duplicate-key errors across DB drivers."""
    message = str(getattr(exc, "orig", exc)).lower()
    return "unique" in message or "duplicate key" in message


@router.get("/", response_model=WorkerListResponse)
async def list_workers(
    service: WorkerService = Depends(get_worker_service),
):
    """List all registered workers."""
    workers = await service.list_workers()
    return WorkerListResponse(workers=workers, total=len(workers))


@router.get("/tasks", response_model=WorkerTaskQueueListResponse)
async def list_worker_tasks(
    active_only: bool = Query(default=True),
    limit: int = Query(default=50, ge=1, le=200),
    service: WorkerService = Depends(get_worker_service),
):
    """List processing/training queue tasks for admin visibility."""
    tasks, total = await service.list_queue_tasks(active_only=active_only, limit=limit)
    return WorkerTaskQueueListResponse(tasks=tasks, total=total)


@router.post("/tasks/cleanup-stale", response_model=WorkerTaskQueueCleanupResponse)
async def cleanup_stale_worker_tasks(
    service: WorkerService = Depends(get_worker_service),
):
    """Run stale claimed-task cleanup immediately."""
    return WorkerTaskQueueCleanupResponse(**await service.cleanup_stale_tasks())


@router.get("/install-info", response_model=WorkerInstallInfo)
async def get_install_info(
    _user: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
):
    """Return package/version + per-platform install targets for the Add-Worker UI.

    Declared before ``/{worker_id}`` so the static path is not captured as an id.
    """
    public_url = await GeneralSettingsService(db).get_public_base_url()
    return WorkerInstallInfo(
        package="intextum-worker",
        version=get_app_version(),
        default_capabilities=_DEFAULT_WORKER_CAPABILITIES,
        public_url=public_url,
        platforms=_WORKER_INSTALL_PLATFORMS,
    )


@router.post(
    "/", response_model=WorkerCreateResponse, status_code=status.HTTP_201_CREATED
)
async def create_worker(
    body: WorkerCreate,
    service: WorkerService = Depends(get_worker_service),
):
    """Create a new worker and generate an API token."""
    try:
        worker = await service.create_worker(
            name=body.name, description=body.description
        )
    except IntegrityError as exc:
        await service.db.rollback()
        if _is_unique_constraint_error(exc):
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=f"Worker with name '{body.name}' already exists",
            )
        raise exc
    token = await service.create_token(worker.id)
    return WorkerCreateResponse(worker=worker, token=token)


@router.get("/{worker_id}", response_model=WorkerResponse)
async def get_worker(
    worker_id: str,
    service: WorkerService = Depends(get_worker_service),
):
    """Get a single worker by ID."""
    worker = await service.get_worker(worker_id)
    if worker is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Worker not found"
        )
    return worker


@router.put("/{worker_id}", response_model=WorkerResponse)
async def update_worker(
    worker_id: str,
    body: WorkerUpdate,
    service: WorkerService = Depends(get_worker_service),
):
    """Update worker name and/or description."""
    try:
        worker = await service.update_worker(
            worker_id, name=body.name, description=body.description
        )
    except IntegrityError as exc:
        await service.db.rollback()
        if _is_unique_constraint_error(exc):
            worker_name = body.name or worker_id
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=f"Worker with name '{worker_name}' already exists",
            )
        raise exc
    if worker is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Worker not found"
        )
    return worker


@router.delete("/{worker_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_worker(
    worker_id: str,
    service: WorkerService = Depends(get_worker_service),
):
    """Delete a worker and revoke its token."""
    deleted = await service.delete_worker(worker_id)
    if not deleted:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Worker not found"
        )
    return None


@router.post("/{worker_id}/rotate-token", response_model=WorkerCreateResponse)
async def rotate_token(
    worker_id: str,
    service: WorkerService = Depends(get_worker_service),
):
    """Rotate a worker's API token. Returns the new token (shown only once)."""
    worker = await service.get_worker(worker_id)
    if worker is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Worker not found"
        )
    token = await service.rotate_token(worker_id)
    return WorkerCreateResponse(worker=worker, token=token)
